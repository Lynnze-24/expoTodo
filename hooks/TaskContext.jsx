import React, { createContext, useContext, useState,useEffect} from 'react'
import uuid from 'react-native-uuid';
import AsyncStorage from '@react-native-async-storage/async-storage';

const TaskContext = createContext();


export function useTaskContext(){
    return useContext(TaskContext)
}

export function TaskProvider({children,mock,val}) {
    // mock and val are for testing
    const initial = mock?mock : [];
    const [task, setTask] = useState(initial);
    const fetchTasks = mock||val? ()=>{return}:async() => {
        let localTasks= await AsyncStorage.getItem('TaskAHL') || null;
        if(localTasks){
            setTask(JSON.parse(localTasks));
        }
    }// second function is for testing 

    
    useEffect(()=> {
        fetchTasks()
    
    }, [])
    
    const addToTask = async(item)=>{
        let newTask = [...task,item]
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    const removeFromTask = async(id)=>{
        let newTask = task.filter(x=> x.id!== id);
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }

    const toggleCheck = async(id) =>{
        let newTask = [...task];
       let changedItem = newTask.find(x=> x.id=== id);
       changedItem.isChecked = !changedItem.isChecked;
       setTask(newTask);
       AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    
    const value = val || {task,addToTask,removeFromTask,toggleCheck};
    
    return (
        <TaskContext.Provider value={value}>
        {children}
      </TaskContext.Provider>
    )
}