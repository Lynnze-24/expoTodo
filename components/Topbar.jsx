import React,{useState} from 'react'
import { View, Text,StyleSheet,Keyboard} from 'react-native'
import { TextInput,IconButton } from 'react-native-paper';
import { useTaskContext } from '../hooks/TaskContext';
import cusTheme from '../theme/Theme';
import uuid from 'react-native-uuid';
// import { Feather } from '@expo/vector-icons';

const Topbar = () => {
    const [text, setText] = useState("");
    const [error, setError] = useState(null);
    const {addToTask} = useTaskContext();
    
    const addNewTask = ()=>{
        if(text===''){
            setError('Please type something')
            return Keyboard.dismiss();
        }
        let newItem = {
            id:uuid.v4(),
            isChecked:false,
            name:text
        }
        addToTask(newItem);
        setText("");
        Keyboard.dismiss()
        if(error){
            setError(null)
        }
        
    }

    const keyPressHandle = function(e){
        if(e.nativeEvent.key === "Enter"){
            addNewTask()
        }
    }



    return (
        <View style={styles.container} >
            <TextInput
            onKeyPress={keyPressHandle }
            dense={true}
            mode='outlined'
            placeholder='Write a todo'
           style={styles.textInput}
            value={text}
            outlineColor={error?cusTheme.danger:cusTheme.primary}
            onChangeText={text => setText(text)}
            activeOutlineColor={error?cusTheme.danger:cusTheme.lpr}
            testID='textInput'
            />
           { error && (<Text testID='errorMessage' style={styles.error}>{error}</Text>)}
             <IconButton
             style={styles.iconButton}
                icon="plus"
                type="MaterialCommunityIcons"
                color={cusTheme.primary}
                size={20}
                onPress={addNewTask}
                testID='plusButton'
            />
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
      backgroundColor:cusTheme.bg,
      padding:20,
      flexDirection:'row',
      justifyContent:'space-between',
      alignItems:'center',
      position:'relative',
    },
    iconButton:{
        width:40,
        height:40,
        marginTop:10,
        marginLeft:10,
        borderColor:cusTheme.primary,
        borderWidth:1,
        borderRadius:20,
        justifyContent:'center',
        alignItems:'center',
    },
    textInput:{
        flex:1,
        backgroundColor:'transparent',
        
    },
    error:{
        position:'absolute',
        top:75,
        left:20,
        color:cusTheme.danger,
    },
});

export default Topbar;
