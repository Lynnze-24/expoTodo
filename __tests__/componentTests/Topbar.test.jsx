import { render, fireEvent } from '@testing-library/react-native';
import Topbar from '../../components/Topbar';
import { TaskProvider } from '../../hooks/TaskContext';

import { Keyboard} from 'react-native'
jest.mock('react-native-uuid');
import uuid from 'react-native-uuid';
import renderer from 'react-test-renderer';
import {describe} from '@jest/globals';

// mock  functions 
    const addToTask = jest.fn();
    const task = jest.fn();
    const removeFromTask = jest.fn();
    const toggleCheck = jest.fn();
    Keyboard.dismiss = jest.fn();


it('renders correctly', () => {
        const comp = renderer.create((<TaskProvider >
            <Topbar/>
        </TaskProvider>)).toJSON();
        expect(comp).toMatchSnapshot();
});



it('should render component correctly',()=>{
    
    const { getByPlaceholderText,getByTestId}  = render(
           <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
               <Topbar/>
           </TaskProvider>
            );

            getByPlaceholderText('Write a todo');
    getByTestId('plusButton');  // will throw error if there is no child components


})


it('should change text correctly when user types',()=>{
    
    const { getByPlaceholderText}  = render(
           <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
               <Topbar/>
           </TaskProvider>
            );

          const textInput = getByPlaceholderText('Write a todo');
          
           fireEvent.changeText(textInput,'go shopping');
           
           expect(textInput).toHaveProp('value','go shopping');
          
})





describe('when user clicks plus button',()=>{
    it('should show error message if no input',()=>{
    
        const { getByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
                const plusBtn = getByTestId('plusButton');
                const textInput = getByTestId('textInput');
                fireEvent.changeText(textInput,'');
                fireEvent.press(plusBtn);
                getByTestId('errorMessage'); //checking error message is there or not
    
    })
    
    it('should create new todo if input is valid and not show error message',()=>{
        
        const { getByTestId,queryByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
                const plusBtn = getByTestId('plusButton');
                const textInput = getByTestId('textInput');
               
                fireEvent.changeText(textInput,'go shopping');
                fireEvent.press(plusBtn);
               
                let newItem = {
                    id:uuid.v4(),
                    isChecked:false,
                    name:'go shopping'
                }
                
                expect(addToTask).toHaveBeenCalledWith(newItem);
                expect(queryByTestId('errorMessage')).toBe(null); //checking error message is there or not
                
    })

    it('should reset textInput and dismiss keyboard if input is valid ',()=>{
    
        const { getByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
                const plusBtn = getByTestId('plusButton');
                const textInput = getByTestId('textInput');
                fireEvent.changeText(textInput,'go shopping');
                fireEvent.press(plusBtn);
                
                expect(textInput).toHaveProp('value','');
                expect(Keyboard.dismiss).toHaveBeenCalled();
    })
    
    
    
    it('should hide error message after successfully creating new todo',()=>{
        
        const { getByTestId,queryByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
                const plusBtn = getByTestId('plusButton');
                const textInput = getByTestId('textInput');
                fireEvent.changeText(textInput,'');
                fireEvent.press(plusBtn);
                getByTestId('errorMessage'); //invalid input so show error message
                fireEvent.changeText(textInput,'go shopping');
                fireEvent.press(plusBtn);
              
                let newItem = {
                    id:uuid.v4(),
                    isChecked:false,
                    name:'go shopping'
                }
                
                expect(addToTask).toHaveBeenCalledWith(newItem);
                expect(queryByTestId('errorMessage')).toBe(null);   //checking error message is there or not
                
    })
})


describe('when user clicks enter key',()=>{
    it('should show error message if no input',()=>{
    
        const { getByTestId,queryByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
                
                const textInput = getByTestId('textInput');
                
                fireEvent.changeText(textInput,'');
                fireEvent(textInput,'submitEditing');
                queryByTestId('errorMessage'); //checking error message is there or not
    
    })
    
    it('should create new todo if input is valid and not show error message',()=>{
        
        const { getByTestId,queryByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
               
                const textInput = getByTestId('textInput');
               
                fireEvent.changeText(textInput,'go shopping');
                fireEvent(textInput,'submitEditing');
               
                let newItem = {
                    id:uuid.v4(),
                    isChecked:false,
                    name:'go shopping'
                }
                
                expect(addToTask).toHaveBeenCalledWith(newItem);
                expect(queryByTestId('errorMessage')).toBe(null); //checking error message is there or not
                
    })

    it('should reset textInput and dismiss keyboard if input is valid ', ()=>{
    
        const { getByTestId,queryByText}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
                
                const textInput = getByTestId('textInput');
                fireEvent.changeText(textInput,'go shopping');
                fireEvent(textInput,'submitEditing');
                expect(queryByText('go shopping')).toBe(null);
                expect(Keyboard.dismiss).toHaveBeenCalled();
    })
    
    
    
    it('should hide error message after successfully creating new todo',()=>{
        
        const { getByTestId,queryByTestId}  = render(
               <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
                   <Topbar/>
               </TaskProvider>
                );
                
                const textInput = getByTestId('textInput');
                fireEvent.changeText(textInput,'');
                fireEvent(textInput,'submitEditing');
                queryByTestId('errorMessage'); //invalid input so show error message
                fireEvent.changeText(textInput,'go shopping');
                fireEvent(textInput,'submitEditing');
              
                let newItem = {
                    id:uuid.v4(),
                    isChecked:false,
                    name:'go shopping'
                }
                
                expect(addToTask).toHaveBeenCalledWith(newItem);
                expect(queryByTestId('errorMessage')).toBe(null);   //checking error message is there or not
                
    })
})








