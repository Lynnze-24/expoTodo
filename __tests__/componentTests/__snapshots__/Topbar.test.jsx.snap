// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fffeee",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "padding": 20,
      "position": "relative",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          Object {
            "backgroundColor": "transparent",
            "borderColor": "#120311",
            "borderRadius": 4,
            "borderWidth": 1,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          Array [
            Object {
              "paddingBottom": 0,
            },
            Object {
              "minHeight": 48,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="Write a todo"
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          selectionColor="rgba(18,3,17,0.6)"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "paddingHorizontal": 14,
                "zIndex": 1,
              },
              Object {
                "height": 40,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              Array [
                Object {},
              ],
            ]
          }
          testID="textInput"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 6,
        "left": 6,
        "right": 6,
        "top": 6,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "margin": 6,
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 15,
            "height": 30,
            "width": 30,
          },
          undefined,
          Object {
            "alignItems": "center",
            "borderColor": "#120311",
            "borderRadius": 20,
            "borderWidth": 1,
            "height": 40,
            "justifyContent": "center",
            "marginLeft": 10,
            "marginTop": 10,
            "width": 40,
          },
        ],
      ]
    }
    testID="plusButton"
  >
    <View>
      <Text />
    </View>
  </View>
</View>
`;
